### PR TITLE
Enable Filter Contributors to access netty event loop while constructing Filters - alternative

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

We want to begin making some resources available to Filter instances at contruction time. Some examples might be:
* executors like the event loop for the filter, or a shared blocking pool
* user-defined expensive components which may have their own threading model and pools
* virtual cluster name
* shared state that we want to be available across channels

This adds a concept of `Context` to `io.kroxylicious.proxy.service`:
```
public interface Context {
    BaseConfig getConfig();
}
```
and the Contributor interface changes to
```
public interface Contributor<T, S extends Context> {
    Class<? extends BaseConfig> getConfigType(String shortName);

    T getInstance(String shortName, S context);
}
```

`Context` is meant to represent information about the context in which the Contributor is being invoked to get an instance. So for a FilterContributor, the invocation will be relative to a virtual cluster and a netty channel. It also contains the configuration object parsed from the YAML configuration file.

We then add a `FilterContext` class used by `FilterContributor` that can offer the netty event loop, plus extend the `BaseContributor` to provide typesafe conveniences for registering filters. Users of BaseContributor will be able

### example usage

given a filter implementation `ExampleFilter` with a custom configuration class `ExampleConfig` and a constructor like:
```
ExampleFilter(ScheduledExecutorService executorService, ExampleConfig config)
````
We can populate it with the event loop in a custom Contributor implementation like this":

```java
import io.kroxylicious.proxy.service.BaseContributor;

public class ExampleContributor extends BaseContributor<KrpcFilter, FilterContext> implements FilterContributor {

    public static final BaseContributorBuilder<KrpcFilter, FilterContext> FILTERS = BaseContributor.<KrpcFilter, FilterContext> builder()
            .add("Example", ExampleConfig.class, (context, config) -> {
                return new ExampleFilter(context.executors().eventLoop(), config);
            });

    public TestFilterContributor() {
        super(FILTERS);
    }
}
```

Or it could accept FilterContext directly like:
```
ExampleFilter(FilterContext context, ExampleConfig config)
````

```java
import io.kroxylicious.proxy.service.BaseContributor;

public class ExampleContributor extends BaseContributor<KrpcFilter, FilterContext> implements FilterContributor {

    public static final BaseContributorBuilder<KrpcFilter, FilterContext> FILTERS = BaseContributor.<KrpcFilter, FilterContext> builder()
            .add("Example", ExampleConfig.class, ExampleFilter::new);

    public TestFilterContributor() {
        super(FILTERS);
    }
}
```

Or if there is no configuration for that filter it could accept:
```
ExampleFilter(FilterContext context)
````

```java
import io.kroxylicious.proxy.service.BaseContributor;

public class ExampleContributor extends BaseContributor<KrpcFilter, FilterContext> implements FilterContributor {

    public static final BaseContributorBuilder<KrpcFilter, FilterContext> FILTERS = BaseContributor.<KrpcFilter, FilterContext> builder()
            .add("Example", ExampleFilter::new);

    public TestFilterContributor() {
        super(FILTERS);
    }
}
```

or

```
ExampleFilter(ScheduledExecutorService service)
````

```java
import io.kroxylicious.proxy.service.BaseContributor;

public class ExampleContributor extends BaseContributor<KrpcFilter, FilterContext> implements FilterContributor {

    public static final BaseContributorBuilder<KrpcFilter, FilterContext> FILTERS = BaseContributor.<KrpcFilter, FilterContext> builder()
            .add("Example", context -> new ExampleFilter(context.executors().eventLoop()));

    public TestFilterContributor() {
        super(FILTERS);
    }
}
```
### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by robobario** [comment]
I think this should supercede #575 as it's nicer that every Contributor's Context does something useful (contain the BaseConfig).

File: integrationtests/src/test/java/io/kroxylicious/proxy/filter/ForwardContext.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+
+public record ForwardContext(KrpcFilterContext filterContext, FilterContext constructionContext, ApiMessage body) {
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/proxy/filter/ForwardContext.java:11]: `ForwardContext` doesn't really mean anything to me. Maybe `ForwardingContext` or `PropagationContext` or even `DelayedConstructionContext`

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Context in which a Contributor is getting an instance, this includes the user-supplied configuration.
+ */
+public interface Context {
+
+    /**
+     * service configuration which may be null if the service instance does not accept configuration.
+     * @return config
+     */
+    BaseConfig getConfig();
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java:20]: Maybe not part of this PR.

Can we add something like `<C extends BaseConfig> C getMyConfig(TypeReference<C> tr)` ? and does it help provide type safe config where we need it?

File: integrationtests/src/test/java/io/kroxylicious/proxy/filter/RejectingCreateTopicFilter.java [hunk 3]
```diff
@@ -27,8 +27,10 @@ public class RejectingCreateTopicFilter implements CreateTopicsRequestFilter {
     public static final String ERROR_MESSAGE = "rejecting all topics";
     private final ForwardingStyle forwardingStyle;
     private final boolean withCloseConnection;
+    private final FilterContext constructionContext;
 
-    public RejectingCreateTopicFilter(RejectingCreateTopicFilterConfig config) {
+    public RejectingCreateTopicFilter(FilterContext constructionContext, RejectingCreateTopicFilterConfig config) {
+        this.constructionContext = constructionContext;
```
[the remark this answers is not in the document]
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/filter/RejectingCreateTopicFilter.java:33]: You mean the filter holding a reference to the construction context. Yeah, we can't really prevent a filter author doing this, the context will be part of the kroxylicious-api and available in the contributor, so there's nothing to prevent them passing it through and holding a reference.

[the remark this answers is not in the document]
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/filter/RejectingCreateTopicFilter.java:33]: I suppose you could make the construction context have a life-cycle and close it after construction, then throw errors if the user kept using it. But that doesn't feel so intuitive either.

[the remark this answers is not in the document]
Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/filter/RejectingCreateTopicFilter.java:33]: Yeah. I don't think we actually need to prevent it as you say it makes sense.

My mental model had it as something ephemeral and short lived but I don't think it needs to be

File: integrationtests/src/test/java/io/kroxylicious/proxy/filter/ForwardContext.java (hunk 1 above)
Comment (robobario) [reply integrationtests/src/test/java/io/kroxylicious/proxy/filter/ForwardContext.java]: renamed to `ForwardingContext` to match `ForwardingStyle`

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java (hunk 2 above)
Comment (robobario) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java:20]: Definitely keen to defer any parameterization of BaseConfig, this is where the wheels have come off a couple of times while I've refactored, so if it can be done in it's own small PR that'd help

**Comment by robobario** [comment]
@SamBarker any more comment on this one?

**Comment by SamBarker** [comment]
> @SamBarker any more comment on this one?

Yeah. I do have a couple of thoughts floating around but need a bit of time to solidify them

**Comment by SamBarker** [comment]
> > @SamBarker any more comment on this one?
> 
> Yeah. I do have a couple of thoughts floating around but need a bit of time to solidify them

Oh wait, I've got my PR's confused. Checking on this one now...

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Context.java:20]: Yeah, ok lets review that one later.

**Review by SamBarker — APPROVED** [review]
Thanks @robobario I definitely think this pushes us forward.
